### PR TITLE
Issue 903/custom panels

### PR DESCRIPTION
**Pull request by raitisbe**

### CODE REVIEW

**Comment by jmacura** [comment]
Great! I will try to look at this tomorrow.

**Comment by raitisbe** [comment]
@jmacura See https://github.com/hslayers/project-rostenice/commits/upgrade-hslayers-2

**Comment by jmacura** [comment]
image

image

Awesome!

**Comment by jmacura** [comment]
It works just charmingly 😋 
image
image
image

File: components/layout/panels/panel-component.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import {ViewRef} from '@angular/core';
+
+export interface HsPanelComponent {
+  viewRef: ViewRef;
+  data: any;
+  isVisible();
```
Comment (jmacura) [inline components/layout/panels/panel-component.ts:6]: add boolean as a required return value type

File: components/layout/panels/panel-item.class.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import {Type} from '@angular/core';
+
+export class HsPanelItem {
+  constructor(public component: Type<any>, public data: any) {}
+}
+
+/*
```
Comment (jmacura) [inline components/layout/panels/panel-item.class.ts:7]: I feel like the following lines can be safely deleted 😉

**Review by jmacura — CHANGES_REQUESTED** [review]
Following our current naming practice, I propose to rename `panel-component.ts` to `panel-component.interface.ts`
and `panel-item.class.ts` to `panel-item.ts`
(reasoning behind that: basically everything is a class now -- all components, services or modules -- but just a few files define interfaces)

**Some general questions:**
Why have you decide to make default panels have `ng-show="panelVisible('name', this)">` defined in the layout.html template, but custom panels must implement the `ng-show` (or `[hidden]`) independently via `isVisible()`?

What is the intended usage of the `data` property in the custom HsPanelComponent?

**Comment by raitisbe** [comment]
> Why have you decide to make default panels have `ng-show="panelVisible('name', this)">` defined in the layout.html template, but custom panels must implement the `ng-show` (or `[hidden]`) independently via `isVisible()`?
> 
Not so much that I did decide, more like I could not find a way how to set the ng-show from outside (from the container component). isVisible in the interface is just so all the components would follow more or less the same principle how to toggle themselves on/off.

**Comment by raitisbe** [comment]
> What is the intended usage of the `data` property in the custom HsPanelComponent?

Pretty much leftover from dialog component, but there are hypothetical use cases, for example having 2 instances of the same panel with various headings or something such.

**Review by jmacura — APPROVED** [review]

**Comment by jmacura** [comment]
> 
> 
> > Why have you decide to make default panels have `ng-show="panelVisible('name', this)">` defined in the layout.html template, but custom panels must implement the `ng-show` (or `[hidden]`) independently via `isVisible()`?
> 
> Not so much that I did decide, more like I could not find a way how to set the ng-show from outside (from the container component). isVisible in the interface is just so all the components would follow more or less the same principle how to toggle themselves on/off.

I see. Good choice then, to define the method in the interface.

**Pull request merged**